libs/ledger/include/ledger/chain/main_chain.hpp

```diff
@@ -128,8 +128,8 @@ class MainChain
   /// @name Block Management
   /// @{
   BlockStatus AddBlock(Block const &block);
-  BlockPtr    GetBlock(BlockHash hash) const;
-  bool        RemoveBlock(BlockHash hash);
+  BlockPtr    GetBlock(BlockHash const &hash) const;
```

Comment:
I would be careful here, this does not feel correct. The internal implementation of this method uses moving semantic on `hash` input parameter. Changing this to reference in method signature effectively results to invalidating content of the external variable passed in by caller of this method.

I'm not completely sure why it was originally implemented this way, but my understanding is that it was intentional to avoid risk that pointer to data which `ConstByteArray` carries to won't get changed, what is critical in multi-threaded scenario in which this method is actually used (e.g. mining thread pool, etc. ...).

Keep in mind that `ConstByteArray` is stores data  using `std::shared_ptr<...>` thus, passing it _by-value_ carriesminimal performance penalty when compared to passing _by-reference_.

This comment applies to all methods bellow with the analogical change.

Reply:
As far as I can tell none of the methods in question are called in a multithreaded context - even `DetectDuplicateTransactions` is called by the miner on a single thread, not the thread pool. The `std::move`s you refer to are no longer in place.

Reply:
The guts of this has been reworked in to state machine (=single thread workflow) since I saw it the last time. Back then methods have been actually called from multiple threads, and that was probably the reason behind passing _by-value_ & using moving semantics (rather than _by-reference_).